File: 3491-find-the-maximum-length-of-valid-subsequence-ii/3491-find-the-maximum-length-of-valid-subsequence-ii.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        from collections import defaultdict
        n=len(nums)
        dp=[defaultdict(int) for i in range(n)]
        max_len=1
        for a in range(n):
            for b in range(a):
                mod=(nums[a]+nums[b])%k
                dp[a][mod] = max(dp[a][mod], dp[b].get(mod, 1) + 1)
                max_len = max(max_len, dp[a][mod])
        return max_len
        

File: 3491-find-the-maximum-length-of-valid-subsequence-ii/3491-find-the-maximum-length-of-valid-subsequence-ii.java
import java.util.*;

class Solution {
    public int maximumLength(int[] nums, int k) {
        int n = nums.length;
        int maxLen = 1;
        Map<Integer, Integer>[] dp = new HashMap[n];
        for (int i = 0; i < n; i++) {
            dp[i] = new HashMap<>();
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                int mod = (nums[j] + nums[i]) % k;
                int prevLen = dp[j].getOrDefault(mod, 1);
                dp[i].put(mod, Math.max(dp[i].getOrDefault(mod, 0), prevLen + 1));
                maxLen = Math.max(maxLen, dp[i].get(mod));
            }
        }

        return maxLen;
    }

}
